### PR TITLE
Add support for GraphQL Schema Language

### PR DESCRIPTION
**Pull request by LegNeato**

### CODE REVIEW

**Comment by LegNeato** [comment]
A couple of notes:

1. Naming is probably bad, please give suggestions if so.
2.  I didn't document the `graphql-parser-integration` feature. Not sure if we want that as public API, though I currently made it public in case other projects need to use it.
3. I had a trait `AsSchemaLanguage` and implemented it for every `MetaType`. I didn't like how it made the code look with all the conditional compilation, so I centralized it into an `AstTranslator` instead.

**Comment by theduke** [comment]
I'm not so sure what the use case for this is.

The schema language is usually used the other way around AFAIK, as in bootstrapping a server from a schema file and adding resolver methods.

Downstream tools usually use a introspection query to get the schema.

Where do you see this being useful?

**Comment by LegNeato** [comment]
You need a schema file to use https://github.com/graphql-rust/graphql-client among other use cases such as external tools. If the schema is defined in Juniper/rust, it is essentially "locked up" until runtime.

**Comment by LegNeato** [comment]
But now that I think about it, perhaps it makes sense for this to be its own little crate/project? An argument can be made that it should be part of juniper as well as not. I guess the same arguments would apply to https://github.com/graphql-rust/juniper/issues/307 as well?

**Comment by theduke** [comment]
First glance looks good, I'll review thoroughly later  today.

One question is how much work it would be to fully switch over to  `graphql-parser` instead of bringing it in as a feature.

Also maybe there is a more elegant  way to detect builting types  than `__` prefix checking.

**Comment by LegNeato** [comment]
@theduke For switching entirely over to `graphql-parser`'s AST I think the concerns in https://github.com/graphql-rust/juniper/issues/138#issuecomment-368644460 are still valid?

**Comment by theduke** [comment]
@LegNeato as a first step I have opened a PR that allows a borrowed AST on the parser: https://github.com/graphql-rust/graphql-parser/pull/19

I'll investigate wheter we'd really want to switch to that AST, and compare it with our own.
But, mapping the AST from the parser to ours would be a very lightweight operation that shouldn't make a difference performance wise.

**Comment by theduke** [comment]
@LegNeato I'm working on switching to graphql-parser.

We could wait with this until that happens, or move ahead now if you'd like to get it in sooner.

File: juniper/src/schema/meta.rs [hunk 1]
```diff
@@ -165,6 +165,14 @@ pub struct Field<'a, S> {
     pub deprecation_status: DeprecationStatus,
 }
 
+impl<'a, S> Field<'a, S> {
+    /// Returns true if the type is built-in to GraphQL.
+    pub fn is_builtin(&self) -> bool {
+        // "used exclusively by GraphQL’s introspection system"
```
Comment (theduke) [inline juniper/src/schema/meta.rs]: This seems a bit hacky, `__` is not reserved so users could add their own objects/ fields/arguments/...

We could add a `builtin: bool` flag to the relevant schema types.

**Comment by theduke** [comment]
Mostly looks good. 

Actually I'd be in favor of including it now, but without the feature, let's just build it in.

**Comment by LegNeato** [comment]
Ok, I'll rip out the features, rebase, and switch to borrowing after your patch lands.

**Comment by arlyon** [comment]
Juniper has changed quite a bit since August (especially with the recommendation to base new PRs on the async branch). That said, keen to test this PR out, I rebased this onto #493 and updated graphql-parser to depend on master and after sprinkling in some lifetimes this works like a charm.

Seems like for this to be merged we are waiting on:

- the `async-await` branch merge
- a new version of `graphql-parser`

I can open a PR with my (minimal) changes, just know that I am still quite new to rust.

**Comment by LegNeato** [comment]
I updated master to use the async-await branch.

**Comment by LegNeato** [comment]
Would love a PR with the updated changes!

File: juniper/src/schema/meta.rs (hunk 1 above)
Comment (LegNeato) [reply juniper/src/schema/meta.rs]: From https://github.com/graphql/graphql-spec/blob/master/spec/Section%203%20--%20Type%20System.md#schema:

"All types and directives defined within a schema must not have a name which begins with {"__"} (two underscores), as this is used exclusively by GraphQL's introspection system."

**Comment by LegNeato** [comment]
@arlyon any chance you can put up the rebased changes?

**Comment by LegNeato** [comment]
Friendly 👉@arlyon

**Comment by arlyon** [comment]
Hey, missed these for some reason. I was going to open a PR, but I noticed a new version of graphql-parser is out and it seems silly not to update. Lifetimes are a bit annoying to iron out so it's taking a bit of time.

**Comment by LegNeato** [comment]
@arlyon great, thank you! Yes, the new version should allow us to use &str instead of Strings and be a lot more memory efficient feel free to post WIPs if you get stuck!

**Comment by LegNeato** [comment]
This was landed as https://github.com/graphql-rust/juniper/pull/676

**Pull request closed without merging**